File: Android/src/app/src/main/java/com/google/ai/edge/gallery/customtasks/stt/SttViewModel.kt
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.ai.edge.gallery.customtasks.stt

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.ai.edge.gallery.customtasks.speech.AudioRecorder
import com.google.ai.edge.gallery.customtasks.speech.SPEECH_SAMPLE_RATE
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * UI state for [SttScreen].
 *
 * @param isRecording Whether the microphone is currently capturing audio.
 * @param isTranscribing Whether captured audio is being decoded.
 * @param transcript The latest transcription result.
 * @param error The latest error message, or empty if none.
 */
data class SttUiState(
  val isRecording: Boolean = false,
  val isTranscribing: Boolean = false,
  val transcript: String = "",
  val error: String = "",
)

@HiltViewModel
class SttViewModel @Inject constructor() : ViewModel() {
  private val _uiState = MutableStateFlow(SttUiState())
  val uiState = _uiState.asStateFlow()

  private val recorder = AudioRecorder(sampleRate = SPEECH_SAMPLE_RATE)

  fun startRecording() {
    try {
      recorder.start()
      _uiState.update { it.copy(isRecording = true, error = "") }
    } catch (e: Throwable) {
      _uiState.update { it.copy(isRecording = false, error = e.message ?: "Failed to start recording") }
    }
  }

  fun stopAndTranscribe(instance: SttModelInstance) {
    if (!uiState.value.isRecording) {
      return
    }
    val samples = recorder.stop()
    _uiState.update { it.copy(isRecording = false, isTranscribing = true) }

    viewModelScope.launch {
      try {
        if (samples.isEmpty()) {
          _uiState.update { it.copy(isTranscribing = false, error = "No audio was recorded") }
          return@launch
        }
        val text =
          withContext(Dispatchers.Default) {
            val stream = instance.recognizer.createStream()
            try {
              stream.acceptWaveform(samples = samples, sampleRate = SPEECH_SAMPLE_RATE)
              instance.recognizer.decode(stream)
              instance.recognizer.getResult(stream).text
            } finally {
              stream.release()
            }
          }
        _uiState.update { it.copy(isTranscribing = false, transcript = text) }
      } catch (e: Throwable) {
        _uiState.update {
          it.copy(isTranscribing = false, error = e.message ?: "Failed to transcribe audio")
        }
      }
    }
  }

  fun cancelRecording() {
    if (uiState.value.isRecording) {
      recorder.stop()
      _uiState.update { it.copy(isRecording = false) }
    }
  }

  override fun onCleared() {
    cancelRecording()
  }
}

File: iOS/Gallery/CustomTasks/Stt/SttViewModel.swift
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License"); see LICENSE.
//
// Port of customtasks/stt/SttViewModel.kt

import AVFoundation
import Combine
import Foundation
import Speech

// MARK: - UI state

/// UI state for `SttScreen`.
struct SttUiState {
  var isRecording: Bool = false
  var isTranscribing: Bool = false
  var transcript: String = ""
  var error: String = ""
}

// MARK: - SttViewModel

@MainActor
final class SttViewModel: ObservableObject {
  @Published var uiState = SttUiState()

  private let recorder = AudioRecorder(sampleRate: SPEECH_SAMPLE_RATE)
  private var transcribeTask: Task<Void, Never>?

  // MARK: - Public API

  func startRecording() {
    do {
      try recorder.start()
      uiState = SttUiState(isRecording: true)
    } catch {
      uiState = SttUiState(error: error.localizedDescription)
    }
  }

  func stopAndTranscribe(instance: SttModelInstance) {
    guard uiState.isRecording else { return }
    let samples = recorder.stop()
    uiState.isRecording = false

    guard !samples.isEmpty else {
      uiState.error = "녹음된 오디오가 없습니다."
      return
    }

    uiState.isTranscribing = true
    transcribeTask?.cancel()
    transcribeTask = Task { [weak self] in
      guard let self else { return }
      do {
        let text = await Task.detached(priority: .userInitiated) {
          await instance.engine.transcribe(samples: samples, sampleRate: SPEECH_SAMPLE_RATE)
        }.value
        await MainActor.run {
          self.uiState.isTranscribing = false
          self.uiState.transcript = text
        }
      }
    }
  }

  func cancelRecording() {
    guard uiState.isRecording else { return }
    _ = recorder.stop()
    uiState.isRecording = false
  }

  /// Requests microphone permission and calls `onGranted` / `onDenied` on the main thread.
  func requestMicrophonePermission(onGranted: @escaping () -> Void, onDenied: @escaping () -> Void) {
    AVAudioApplication.requestRecordPermission { granted in
      DispatchQueue.main.async {
        if granted { onGranted() } else { onDenied() }
      }
    }
  }

  deinit {
    // Note: `deinit` is not isolated; cancel the task synchronously.
    transcribeTask?.cancel()
  }
}
